hazelcast/src/test/java/com/hazelcast/spi/impl/executionservice/impl/BasicCompletableFutureTest.java

```diff
@@ -0,0 +1,275 @@
+
+        assertEquals(DELEGATE_RESULT, outerFuture.get(10, TimeUnit.MILLISECONDS));
+    }
+
+    @Test
+    // PROBLEM NR.2
+    // Callbacks do not run if delegate future isDone but outer is not
+    public void completeDelegate_successfully_callbacksNeverRun() throws Exception {
+        ExecutionCallback callback = mock(ExecutionCallback.class);
+
+        delegateFuture.run();
+        outerFuture.andThen(callback);
+
+        verifyZeroInteractions(callback);
+    }
+
+    @Test
+    // PROBLEM NR.2
+    public void completeDelegate_withException_callbacksNeverRun() throws Exception {
+        ExecutionCallback callback = mock(ExecutionCallback.class);
+        delegateThrowException = true;
+
+        delegateFuture.run();
+        outerFuture.andThen(callback);
+
+        verifyZeroInteractions(callback);
+    }
+
+    @Test
+    public void completeDelegate_successfully_invokeIsDoneOnOuter_callbacksRun() throws Exception {
+        ExecutionCallback callback = mock(ExecutionCallback.class);
+
+        delegateFuture.run();
+        outerFuture.isDone();
+        outerFuture.andThen(callback);
+
+        verify(callback, times(1)).onResponse(any(Throwable.class));
+        verify(callback, times(0)).onFailure(any(Throwable.class));
+    }
+
+    @Test
+    public void completeDelegate_successfully_invokeGetOnOuter_callbacksRun() throws Exception {
+        ExecutionCallback callback = mock(ExecutionCallback.class);
+
+        delegateFuture.run();
+        outerFuture.get();
+        outerFuture.andThen(callback);
+
+        verify(callback, times(1)).onResponse(any(Throwable.class));
+        verify(callback, times(0)).onFailure(any(Throwable.class));
+    }
+
+    @Test
+    // PROBLEM NR.3
+    // Inconsistent behavior of callbacks on exception - if Delegate future finishes with an exception
+    // and isDone() is called on BCF the callback is called with onFailure()!
+    public void completeDelegate_withException_invokeIsDoneOnOuter_callbacksRun() throws Exception {
+        ExecutionCallback callback = mock(ExecutionCallback.class);
+        delegateThrowException = true;
+
+        delegateFuture.run();
+        outerFuture.isDone();
+        outerFuture.andThen(callback);
+
+        verify(callback, times(0)).onResponse(any(Throwable.class));
+        verify(callback, times(1)).onFailure(any(Throwable.class));
+    }
+
+    @Test
+    // PROBLEM NR.3
+    // ...however, if the Delegate future finishes with an exception
+    // and get() is called on BCF the callbacks will not run at all!
+    public void completeDelegate_withException_invokeGetOnOuter_callbacksNeverReached() throws Exception {
+        ExecutionCallback callback = mock(ExecutionCallback.class);
+        delegateThrowException = true;
+
+        delegateFuture.run();
+
+        expected.expectMessage("Exception in execution");
+        outerFuture.get();
+
+        // callbacks will be never reached
+        outerFuture.andThen(callback);
+
```

Comment:
Yes, that's true - that's the reason I've put this comment `// callbacks will be never reached`
But I agree it may be confusing - I'll remove everything below `outerFuture.get()` and just leave a one line of comment there.